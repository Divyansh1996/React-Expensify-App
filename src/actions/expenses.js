import {v4 as uuid4} from 'uuid';
import db from '../firebase/firebase'
import { push,ref,get,remove, update} from 'firebase/database';

export const addExpense = (expense) => (
    {
        type:'ADD_EXPENSE',
        expense
    })

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {description = '', note = '', amount = 0, createdAt = ''} = expenseData
        const expense = {description, note, amount, createdAt}
        return push(ref(db, `users/${uid}/expenses`), expense).then((ref) => {
            dispatch(addExpense({id:ref.key, ...expense}))
        })

    }
}

export const removeExpense=({id=''}={}) => ({
    type:'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense =({id}) => {
    return(dispatch, getState) => {
        const uid = getState().auth.uid;
        return remove(ref(db, `users/${uid}/expenses/${id}`)).then(()=>{
            dispatch(removeExpense({id}))
        })
    }

}


export const editExpense = (id, updates) => ({
    id,
    type:'EDIT_EXPENSE',
    updates
})

export const startEditExpense =(id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return update(ref(db, `users/${uid}/expenses/${id}`),updates).then(() => {
            dispatch(editExpense(id, updates))
        })
    }
}

export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
})

export const startSetExpenses = () =>{
    return(dispatch, getState) => {
        const uid = getState().auth.uid;
        return get(ref(db, `users/${uid}/expenses`)).then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}