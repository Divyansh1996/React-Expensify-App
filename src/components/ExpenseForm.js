import moment from "moment";
import React from "react";
import 'react-dates/initialize';
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component{
    state ={
        description: this.props.expense? this.props.expense.description : '',
        note: this.props.expense? this.props.expense.note : '',
        amount: this.props.expense? (this.props.expense.amount).toString() : '',
        createdAt: this.props.expense? moment(this.props.expense.createdAt) : moment(),
        calenderFocused: false,
        error:undefined
    }

    OnDescriptionChange = (e) =>{
        const description = e.target.value
        this.setState(()=> ({description}))
    }

    OnNoteChange = (e) =>{
        const note = e.target.value
        this.setState(()=> ({note}))
    }

    OnAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }
    }

    onDateChangeHere = (date) => {
        if(date){
            this.setState(()=>({createdAt:date}))
        }
    }

    onFocusChange=({focused}) => {
        this.setState(()=>({calenderFocused:focused}))
    }

    onSubmitChange = (e) => {
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error: "Please provide the description and amount"}))
        }
        else{
            this.setState(()=>({error: undefined}))
            this.props.onSubmit(({
                description: this.state.description,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf(),
                amount:parseFloat(this.state.amount,10)
            }))
        }
    }

    render(){
        return(
            <div>
                <h1>ExpenseForm</h1>
                <h2>{this.state.error}</h2>
                <form onSubmit={this.onSubmitChange}>
                    <input type="text" placeholder="Description" value = {this.state.description} autoFocus onChange={this.OnDescriptionChange}/>
                    <input type="text" placeholder="Amount" value = {this.state.amount} onChange={this.OnAmountChange}/>
                    <textarea placeholder="Add a Note" value = {this.state.note} onChange={this.OnNoteChange} optional></textarea>
                    <SingleDatePicker 
                        date ={this.state.createdAt} 
                        onDateChange={this.onDateChangeHere} 
                        focused = {this.state.calenderFocused} 
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>(false)}
                        />
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}