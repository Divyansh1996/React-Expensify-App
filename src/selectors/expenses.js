import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const momentCreatedAt = moment(expense.createdAt);
        const startMatch = startDate? startDate.isSameOrBefore(momentCreatedAt):true
        const endMatch = endDate? endDate.isSameOrAfter(momentCreatedAt):true
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());
        
        return startMatch && endMatch && textMatch;
    }).sort((a,b) => {
        if(sortBy === 'date'){
            return a.createdAt <=b.createdAt?1:-1
        }
        if(sortBy === 'amount'){
            return a.amount <=b.amount?1:-1
        }
    })
}
