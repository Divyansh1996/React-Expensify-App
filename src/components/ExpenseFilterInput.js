import React from "react"
import { connect } from "react-redux"
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';
import {editText,sortByDate,sortByAmount, setStartDate, setEndDate} from "../actions/filter"
import {DateRangePicker} from 'react-dates'

export class ExpenseFilterInput extends React.Component{
    state={
        calenderFocused:null
    }
    onDatesChange = ({startDate, endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange =(focusedInput) => {
        this.setState(()=>({calenderFocused: focusedInput}))
    }
    onTextChange = (e) => {
        this.props.editText(e.target.value)
    }
    onSortByChange = (e) => {
        e.target.value === "date"? this.props.sortByDate(): this.props.sortByAmount()
    }
    render(){
        return (
            <div>
                <input type="text" value={this.props.filter.text} onChange={this.onTextChange}/>
                <select value={this.props.filter.sortBy} 
                    onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDateId="your_unique_start_date_id"
                    endDateId="your_unique_end_date_id"
                    startDate={this.props.filter.startDate}
                    endDate={this.props.filter.endDate}
                    onDatesChange={this.onDatesChange}    
                    focusedInput={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>(false)}
                    showClearDates={true}
                />
            </div>
        )
    }
}

const mapStatetoProps = (state) => ({
        filter: state.filter
    }
)
const mapDispatchtoProps = (dispatch) => ({
    setStartDate: (startDate) =>{dispatch(setStartDate(startDate))},
    setEndDate: (endDate) =>{dispatch(setEndDate(endDate))},
    editText: (text) => {dispatch(editText(text))},
    sortByDate: () =>{dispatch(sortByDate())},
    sortByAmount: () =>{dispatch(sortByAmount())}
})
export default connect(mapStatetoProps, mapDispatchtoProps)(ExpenseFilterInput)