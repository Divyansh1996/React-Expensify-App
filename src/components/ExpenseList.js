import { connect } from "react-redux";
import React from "react";
import expenses from "../reducers/expenses";
import filter from "../reducers/filter";
import ExpenseListItem from "./ExpenseListItem"
import SelectorExpense from "../selectors/expenses"

export const ExpenseList = (props) => {
    return (<div>
        <h1>ExpenseList</h1>
        {props.expenses.length ==0 && <p>No Expense Found</p>}
        {props.expenses.length > 0 && props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense}/>
        })}
    </div>
)}

const mapStatetoProps = (state) => (
    {
        expenses: SelectorExpense(state.expenses, state.filter)
    }
)
export default connect(mapStatetoProps)(ExpenseList)