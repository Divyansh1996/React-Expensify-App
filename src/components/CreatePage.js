import React from "react";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";
import { connect } from "react-redux";

export class CreatePage extends React.Component{
    onSubmit = (expense) => {
        this.props.addNewExpense(expense)
        this.props.history.push("/")
    }
    render(){
        return (
        <div>
            <ExpenseForm onSubmit={this.onSubmit}/>
        </div>)
    }
}

const mapDispatchtoProps = (dispatch) => ({
        addNewExpense: (expense) => {dispatch(addExpense(expense))}
    }
)

export default connect(undefined,mapDispatchtoProps)(CreatePage);