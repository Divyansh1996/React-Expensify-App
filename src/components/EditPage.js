import React from "react";
import { useParams,useHistory } from "react-router-dom";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import expenses from "../selectors/expenses";
import { editExpense,removeExpense } from "../actions/expenses";

export class EditPage extends React.Component{
    onSubmitChange = (expense) =>{
         this.props.editNewExpense(expense)
         this.props.history.push("/")
    }
    onRemove = () => {
        this.props.removeNewExpense()
        this.props.history.push("/")
    }
    render(){
        return (
            <div>
                <ExpenseForm expense = {this.props.expense} onSubmit={this.onSubmitChange} />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        )
    }
}

const mapStatetoProps = (state,props) => ({
    expense: state.expenses.find((expense) =>expense.id ===props.match.params.id),
})

const mapDispatchtoProps = (dispatch,props) => ({
    editNewExpense: (expense) => {dispatch(editExpense(props.match.params.id,expense))},
    removeNewExpense: () => {dispatch(removeExpense({id:props.match.params.id}))}
})
export default connect(mapStatetoProps, mapDispatchtoProps)(EditPage);