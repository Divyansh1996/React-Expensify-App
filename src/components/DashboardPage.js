import React from "react";

import ExpenseList from "./ExpenseList";
import ExpenseFilterInput from "./ExpenseFilterInput";

const DashboardPage = () => (
    <div>
        <ExpenseFilterInput />
        <ExpenseList />
    </div>
)


export default DashboardPage;