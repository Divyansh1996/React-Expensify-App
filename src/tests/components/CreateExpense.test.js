import React from "react";
import {CreatePage} from "../../components/CreatePage";
import {shallow} from "enzyme";
import expenses from "../fixtures/expenses";

let addNewExpense, history, wrapper

beforeEach(()=> {
    addNewExpense = jest.fn();
    history ={
        push: jest.fn()
    }
    wrapper = shallow(<CreatePage addNewExpense = {addNewExpense}  history = {history}/>)
})

test("Testing the method of CreatePage with default values", () => {
    expect(wrapper).toMatchSnapshot();
})


test("Testing the method of CreatePage with some values", () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(addNewExpense).toHaveBeenCalledWith(expenses[1])
    expect(history.push).toHaveBeenCalledWith('/')
})

