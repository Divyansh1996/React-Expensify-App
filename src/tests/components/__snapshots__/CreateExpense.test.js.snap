// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Testing the method of CreatePage with default values 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
